Add add_tostr flag to Cleaner to control ToStr transformation

PilliSiddharth opened the pull request:
# Add `add_tostr` flag to `Cleaner` to control `ToStr` transformation

## Summary
This PR introduces an `add_tostr` parameter to the `Cleaner` transformer, allowing users to disable the automatic `ToStr()` conversion applied to non-numeric, non-categorical, and non-datetime columns. This prevents list/object-like columns from being converted to strings unless explicitly requested.

## What’s changed
- Add `add_tostr` argument to `Cleaner` (default: `True`).
- Update `_get_preprocessors` to include `ToStr()` only when `add_tostr=True`.
- Update `Cleaner` docstring (parameter documented; notes updated).
- Add `test_add_tostr` to validate both `add_tostr=True` and `add_tostr=False`.

## Motivation
`Cleaner` should not silently change complex dtypes (e.g., `list`, `object`) to strings. This flag gives users control and mirrors the existing pattern used for `numeric_dtype="float32"`.

## Test coverage
- New unit test `skrub/tests/test_cleaner.py::test_add_tostr` added.
- Local test run: relevant tests pass.

rcap107 commented on skrub/_table_vectorizer.py:
```diff
@@ -134,10 +135,11 @@ def _get_preprocessors(
+    if add_tostr:
+        transformers.append(ToStr())
+
+    transformers.append(CleanCategories())
```
The order of the transformers is important. `CleanCategories` should go before `ToStr`, so this part needs to be fixed.

rcap107 commented on skrub/_table_vectorizer.py:
```diff
@@ -243,8 +252,11 @@ class Cleaner(TransformerMixin, BaseEstimator):
-      categorical, or datetime.
+    - ``ToStr()``: convert columns to strings unless they are numerical,
+    categorical, or datetime. This step is controlled by the ``add_tostr``
+    parameter. When ``add_tostr=True`` (default), string conversion is applied.
```
as mentioned above, the name `add_tostr` will probably be changed

rcap107 reviewed (dismissed):
Thanks for the PR @PilliSiddharth. Overall, it looks good, but there are a few things that should be changed.

We should also rethink the name of the `add_tostr` parameter, but at the moment I don't have a very good alternative in mind.

PilliSiddharth commented:
Hey @rcap107, I moved the `test_add_tostr` test into `tests/test_table_vectorizer.py`, set the default to False, fixed the ordering so CleanCategories runs before ToStr, and fixed the docstring text. About the name, `add_tostr` works technically, but I agree it doesn’t feel very descriptive. Something like `convert_objects` might be clearer, but happy to go with whatever you prefer. Let me know if this looks right or if you want me to adjust anything else before I update the PR again.

rcap107 commented on skrub/_table_vectorizer.py:
```diff
@@ -313,13 +323,15 @@ def __init__(
         self.datetime_format = datetime_format
         self.numeric_dtype = numeric_dtype
         self.n_jobs = n_jobs
+        self.cast_to_str = cast_to_str
```
```suggestion
        self.cast_to_str = cast_to_str
        self.n_jobs = n_jobs
```

rcap107 commented on skrub/_table_vectorizer.py:
```diff
@@ -313,13 +323,15 @@ def __init__(
         datetime_format=None,
         numeric_dtype=None,
         n_jobs=1,
+        cast_to_str=False,
```
```suggestion
        cast_to_str=False,
        n_jobs=1,
```

rcap107 commented:
Whoops, the order change broke the CI, I'll fix it

PilliSiddharth commented:
Sounds good, glad I could be a part! 😄
